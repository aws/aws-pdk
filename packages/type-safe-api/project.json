{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
