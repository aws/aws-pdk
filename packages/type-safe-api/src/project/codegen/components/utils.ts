/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import * as readPkg from "read-pkg-up";
import { Language, Library } from "../../languages";
import { MockResponseDataGenerationOptions } from "../../types";

/**
 * Enum for generator directories for non-runtime generators
 */
export enum OtherGenerators {
  DOCS = "docs",
  // Infrastructure
  TYPESCRIPT_CDK_INFRASTRUCTURE = "typescript-cdk-infrastructure",
  PYTHON_CDK_INFRASTRUCTURE = "python-cdk-infrastructure",
  JAVA_CDK_INFRASTRUCTURE = "java-cdk-infrastructure",
  // Handlers
  TYPESCRIPT_LAMBDA_HANDLERS = "typescript-lambda-handlers",
  PYTHON_LAMBDA_HANDLERS = "python-lambda-handlers",
  JAVA_LAMBDA_HANDLERS = "java-lambda-handlers",
}

export enum TypeSafeApiScript {
  PARSE_OPENAPI_SPEC = "parse-openapi-spec",
  GENERATE = "generate",
  GENERATE_MOCK_DATA = "generate-mock-data",
  GENERATE_HTML_REDOC_DOCS = "generate-html-redoc-docs",
  CLEAN_OPENAPI_GENERATED_CODE = "clean-openapi-generated-code",
  COPY_GRADLE_WRAPPER = "copy-gradle-wrapper",
}

/**
 * Generator directory for openapi generation containing templates, config etc.
 */
export type GeneratorDirectory = Language | Library | OtherGenerators;

/**
 * Types of normalizers supported by openapi-generator
 * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
 */
export type OpenApiNormalizer = "KEEP_ONLY_FIRST_TAG_IN_OPERATION";

/**
 * Options for generating code or docs using OpenAPI Generator CLI
 */
export interface GenerationOptions {
  /**
   * The OpenAPI generator to use to generate the code/docs
   */
  readonly generator: string;
  /**
   * The directory to use for OpenAPI generation
   */
  readonly generatorDirectory: GeneratorDirectory;
  /**
   * The path of the OpenAPI spec to generate code from
   */
  readonly specPath: string;
  /**
   * Additional properties to pass to the generate cli
   */
  readonly additionalProperties?: {
    [key: string]: string;
  };
  /**
   * Supply the relative path from the code project root to the source code directory in which custom generated files
   * (eg. operation config) should be placed.
   */
  readonly srcDir?: string;
  /**
   * Normalizers to apply to the spec prior to generation, if any
   * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
   */
  readonly normalizers?: Partial<Record<OpenApiNormalizer, boolean>>;
  /**
   * Vendor extensions to add for code generation, allowing custom properties to be passed to the generator templates
   * Keys should begin with "x-"
   */
  readonly extraVendorExtensions?: Record<string, string | boolean>;
  /**
   * Generate alias as model.
   * Defines whether "primitive types" defined at the model/schema level will be wrapped in a model (ie maps/lists)
   * @see https://openapi-generator.tech/docs/globals/
   * @default true
   */
  readonly generateAliasAsModel?: boolean;
  /**
   * The path to the json smithy model file, if available
   */
  readonly smithyJsonPath?: string;
}

/**
 * Build a command for running a script from this project's bin
 */
export const buildTypeSafeApiExecCommand = (
  script: TypeSafeApiScript,
  args?: string
) => {
  const { packageJson } = readPkg.sync({
    cwd: path.resolve(__dirname),
  })!;
  return `npx --yes -p aws-pdk@${packageJson.version} ${script}${
    args ? ` ${args}` : ""
  }`;
};

const serializeProperties = (properties: { [key: string]: string }) =>
  Object.entries(properties)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");

/**
 * Generate code or docs by invoking the root generate script
 */
export const buildInvokeOpenApiGeneratorCommandArgs = (
  options: GenerationOptions
): string => {
  const srcDir = options.srcDir ?? "src";
  const additionalProperties = options.additionalProperties
    ? ` --additional-properties "${serializeProperties(
        options.additionalProperties
      )}"`
    : "";

  const normalizers = options.normalizers
    ? ` --openapi-normalizer "${serializeProperties(
        Object.fromEntries(
          Object.entries(options.normalizers).map(([k, v]) => [k, `${v}`])
        )
      )}"`
    : "";

  const extensions = options.extraVendorExtensions
    ? ` --extra-vendor-extensions '${JSON.stringify(
        options.extraVendorExtensions
      )}'`
    : "";

  const generateAliasAsModel =
    options.generateAliasAsModel ?? true ? " --generate-alias-as-model" : "";

  const smithyJsonPath = options.smithyJsonPath
    ? ` --smithy-json-path ${options.smithyJsonPath}`
    : "";

  const specPath = options.specPath;
  const outputPath = ".";

  return `--generator ${options.generator} --spec-path ${specPath} --output-path ${outputPath} --generator-dir ${options.generatorDirectory} --src-dir ${srcDir}${smithyJsonPath}${additionalProperties}${normalizers}${extensions}${generateAliasAsModel}`;
};

/**
 * Builds a command to clean up files which were previously generated by openapi generator
 */
export const buildCleanOpenApiGeneratedCodeCommand = (): string => {
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.CLEAN_OPENAPI_GENERATED_CODE,
    `--code-path .`
  );
};

/**
 * Options for generating mock data json files
 */
export interface MockDataGenerationOptions
  extends MockResponseDataGenerationOptions {
  /**
   * The path of the OpenAPI spec to generate data for
   */
  readonly specPath: string;
  /**
   * Output sub directory relative to the outdir in which to generate mock data
   * Mock data will be written to a directory named 'mocks' within the sub directory
   * @default .
   */
  readonly outputSubDir?: string;
}

/**
 * Invoke the mock data generator script
 */
export const buildInvokeMockDataGeneratorCommand = (
  options: MockDataGenerationOptions
): string => {
  const outputPath = options.outputSubDir ?? ".";
  const locale = options.locale ? ` --locale ${options.locale}` : "";
  const maxArrayLength =
    options.maxArrayLength !== undefined
      ? ` --max-array-length ${options.maxArrayLength}`
      : "";
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.GENERATE_MOCK_DATA,
    `--spec-path ${options.specPath} --output-path ${outputPath}${locale}${maxArrayLength}`
  );
};
