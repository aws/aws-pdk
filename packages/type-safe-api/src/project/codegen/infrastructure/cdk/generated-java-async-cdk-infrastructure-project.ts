/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedJavaCdkInfrastructureBaseProject,
  GeneratedJavaCdkInfrastructureBaseProjectOptions,
} from "./generated-java-cdk-infrastructure-base-project";
import {
  GenerationOptions,
  OtherGenerators,
  getHandlersProjectVendorExtensions,
} from "../../components/utils";

export interface GeneratedJavaAsyncCdkInfrastructureProjectOptions
  extends GeneratedJavaCdkInfrastructureBaseProjectOptions {}

export class GeneratedJavaAsyncCdkInfrastructureProject extends GeneratedJavaCdkInfrastructureBaseProject {
  constructor(options: GeneratedJavaAsyncCdkInfrastructureProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "java",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.JAVA_ASYNC_CDK_INFRASTRUCTURE,
      srcDir: this.srcDir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-infrastructure-package": this.packageName,
        "x-runtime-package": this.options.generatedJavaTypes.packageName,
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    };
  }
}
