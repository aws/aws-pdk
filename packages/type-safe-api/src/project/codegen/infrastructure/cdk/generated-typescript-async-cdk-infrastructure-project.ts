/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import {
  GeneratedTypescriptCdkInfrastructureBaseProject,
  GeneratedTypescriptCdkInfrastructureBaseProjectOptions,
} from "./generated-typescript-cdk-infrastructure-base-project";
import {
  CodegenOptions,
  getHandlersProjectVendorExtensions,
  OtherGenerators,
} from "../../components/utils";

export interface GeneratedTypescriptAsyncCdkInfrastructureProjectOptions
  extends GeneratedTypescriptCdkInfrastructureBaseProjectOptions {}

export class GeneratedTypescriptAsyncCdkInfrastructureProject extends GeneratedTypescriptCdkInfrastructureBaseProject {
  constructor(
    options: GeneratedTypescriptAsyncCdkInfrastructureProjectOptions
  ) {
    super(options);
  }

  public buildCodegenOptions(): CodegenOptions {
    return {
      specPath: this.options.specPath,
      templateDirs: [OtherGenerators.TYPESCRIPT_ASYNC_CDK_INFRASTRUCTURE],
      metadata: {
        srcDir: this.srcdir,
        runtimePackageName:
          this.options.generatedTypescriptTypes.package.packageName,
        // Spec path relative to the source directory
        relativeSpecPath: path.join("..", this.packagedSpecPath),
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
    };
  }
}
