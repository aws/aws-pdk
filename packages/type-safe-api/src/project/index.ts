/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
