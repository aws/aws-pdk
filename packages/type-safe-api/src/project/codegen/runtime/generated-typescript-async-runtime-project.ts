/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedTypescriptRuntimeBaseProject,
  GeneratedTypescriptRuntimeBaseProjectOptions,
} from "./generated-typescript-runtime-base-project";
import { GenerationOptions, OtherGenerators } from "../components/utils";

/**
 * Configuration for the generated typescript client project
 */
export interface GeneratedTypescriptAsyncRuntimeProjectOptions
  extends GeneratedTypescriptRuntimeBaseProjectOptions {}

/**
 * Typescript project containing types generated using OpenAPI Generator CLI
 */
export class GeneratedTypescriptAsyncRuntimeProject extends GeneratedTypescriptRuntimeBaseProject {
  constructor(options: GeneratedTypescriptAsyncRuntimeProjectOptions) {
    super(options);

    this.addDeps("@aws-sdk/client-apigatewaymanagementapi");

    this.openapiGeneratorIgnore.addPatterns(
      // Skip generating http clients
      `${this.srcdir}/apis/**/*`,
      `${this.srcdir}/apis/*`
    );
  }

  protected buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "typescript-fetch",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.TYPESCRIPT_ASYNC_RUNTIME,
      additionalProperties: {
        npmName: this.package.packageName,
        typescriptThreePlus: "true",
        useSingleParameter: "true",
        supportsES6: "true",
      },
      srcDir: this.srcdir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    };
  }
}
