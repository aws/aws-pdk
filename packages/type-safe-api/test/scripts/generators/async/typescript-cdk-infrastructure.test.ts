/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { getTestHandlerProjects } from "./utils";
import { GeneratedTypescriptAsyncCdkInfrastructureProject } from "../../../../src/project/codegen/infrastructure/cdk/generated-typescript-async-cdk-infrastructure-project";
import { GeneratedTypescriptAsyncRuntimeProject } from "../../../../src/project/codegen/runtime/generated-typescript-async-runtime-project";
import { withTmpDirSnapshot } from "../../../project/snapshot-utils";

describe("Typescript Async Infrastructure Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(
      __dirname,
      `../../../resources/specs/async/${spec}`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedTypescriptAsyncRuntimeProject({
        name: "test-client",
        defaultReleaseBranch: "main",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedTypescriptAsyncCdkInfrastructureProject({
        name: "test-infra",
        defaultReleaseBranch: "main",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedTypescriptTypes: client,
        generatedHandlers: {},
      });
      exec(`mkdir -p ${infraOutdir}`, { cwd: outdir });
      exec(
        `${path.resolve(
          __dirname,
          "../../../../scripts/type-safe-api/run.js generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/src/api.ts"]).toMatchSnapshot();
    expect(snapshot["infra/src/mock-integrations.ts"]).toMatchSnapshot();
    expect(snapshot["infra/src/index.ts"]).toMatchSnapshot();
  });

  it.each(["handlers.yaml", "inline-body.yaml"])(
    "Generates Functions for %s",
    (specFile) => {
      const specPath = path.resolve(
        __dirname,
        `../../../resources/specs/async/${specFile}`
      );

      const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
        exec(`cp ${specPath} ${outdir}/spec.yaml`, {
          cwd: path.resolve(__dirname),
        });
        const { runtimes, handlers } = getTestHandlerProjects(outdir);
        const infraOutdir = path.join(outdir, "infra");
        const project = new GeneratedTypescriptAsyncCdkInfrastructureProject({
          name: "test-infra",
          defaultReleaseBranch: "main",
          outdir: infraOutdir,
          specPath: "../spec.yaml",
          generatedTypescriptTypes: runtimes.typescript,
          generatedHandlers: handlers,
        });
        exec(`mkdir -p ${infraOutdir}`, { cwd: outdir });
        exec(
          `${path.resolve(
            __dirname,
            "../../../../scripts/type-safe-api/run.js generate"
          )} ${project.buildGenerateCommandArgs()}`,
          {
            cwd: infraOutdir,
          }
        );
      });

      expect(snapshot["infra/src/functions.ts"]).toMatchSnapshot();
    }
  );
});
