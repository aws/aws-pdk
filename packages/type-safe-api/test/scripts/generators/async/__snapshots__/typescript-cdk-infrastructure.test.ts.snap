// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript Async Infrastructure Code Generation Script Unit Tests Generates Functions for handlers.yaml 1`] = `
"import { Construct } from "constructs";
import { Duration } from "aws-cdk-lib";
import { SnapStartFunction, SnapStartFunctionProps } from "@aws/pdk/type-safe-api";
import { Code, Function, Runtime, Tracing, FunctionProps } from "aws-cdk-lib/aws-lambda";
import * as path from "path";

/**
 * Options for the $ConnectFunction construct
 */
export interface $ConnectFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation for the websocket connect event
 */
export class $ConnectFunction extends Function {
  constructor(scope: Construct, id: string, props?: $ConnectFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "$connect",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the $DisconnectFunction construct
 */
export interface $DisconnectFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation for the websocket disconnect event
 */
export class $DisconnectFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: $DisconnectFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.$DisconnectHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the JavaOneFunction construct
 */
export interface JavaOneFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of JavaOne
 */
export class JavaOneFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaOneHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the JavaTwoFunction construct
 */
export interface JavaTwoFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of JavaTwo
 */
export class JavaTwoFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaTwoHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the PythonOneFunction construct
 */
export interface PythonOneFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of PythonOne
 */
export class PythonOneFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python_one.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the PythonTwoFunction construct
 */
export interface PythonTwoFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of PythonTwo
 */
export class PythonTwoFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python_two.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the TypescriptOneFunction construct
 */
export interface TypescriptOneFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypescriptOne
 */
export class TypescriptOneFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypescriptOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "typescript-one",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the TypescriptTwoFunction construct
 */
export interface TypescriptTwoFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypescriptTwo
 */
export class TypescriptTwoFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypescriptTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "typescript-two",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

"
`;

exports[`Typescript Async Infrastructure Code Generation Script Unit Tests Generates Functions for inline-body.yaml 1`] = `
"import { Construct } from "constructs";
import { Duration } from "aws-cdk-lib";
import { SnapStartFunction, SnapStartFunctionProps } from "@aws/pdk/type-safe-api";
import { Code, Function, Runtime, Tracing, FunctionProps } from "aws-cdk-lib/aws-lambda";
import * as path from "path";



/**
 * Options for the JavaFunction construct
 */
export interface JavaFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of Java
 */
export class JavaFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the PythonFunction construct
 */
export interface PythonFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of Python
 */
export class PythonFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}


/**
 * Options for the TypeScriptFunction construct
 */
export interface TypeScriptFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypeScript
 */
export class TypeScriptFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypeScriptFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "type-script",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

"
`;

exports[`Typescript Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 1`] = `
"import { TypeSafeWebsocketApi, TypeSafeWebsocketApiProps, TypeSafeWebsocketApiIntegration } from "@aws/pdk/type-safe-api";
import { Construct } from "constructs";
import { OperationConfig, OperationLookup } from "test-client";
import * as path from "path";

export type WebSocketApiIntegrations = OperationConfig<TypeSafeWebsocketApiIntegration>;

export interface WebSocketApiProps extends Omit<TypeSafeWebsocketApiProps, "specPath" | "integrations" | "operationLookup"> {
  readonly integrations: WebSocketApiIntegrations;
}

/**
 * Type-safe construct for the API Gateway resources defined by your model.
 * This construct is generated and should not be modified.
 */
export class WebSocketApi extends TypeSafeWebsocketApi {
  constructor(scope: Construct, id: string, props: WebSocketApiProps) {
    super(scope, id, {
      ...props,
      integrations: props.integrations as any,
      operationLookup: OperationLookup,
      specPath: path.resolve(__dirname, "../assets/api.json"),
    });
  }
}
"
`;

exports[`Typescript Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 2`] = `
"import {
  WebSocketMockIntegration,
} from "aws-cdk-lib/aws-apigatewayv2-integrations";


/**
 * Type-safe mock integrations for WebSocket API operations
 */
export class MockIntegrations {

  /**
   * Mock all operations
   */
  public static mockAll() {
    return {
      anyRequest: {
        integration: new WebSocketMockIntegration("MockAnyRequestIntegration"),
      },
      bidirectional: {
        integration: new WebSocketMockIntegration("MockBidirectionalIntegration"),
      },
      clientToServer: {
        integration: new WebSocketMockIntegration("MockClientToServerIntegration"),
      },
      empty: {
        integration: new WebSocketMockIntegration("MockEmptyIntegration"),
      },
      mapRequest: {
        integration: new WebSocketMockIntegration("MockMapRequestIntegration"),
      },
      mediaTypes: {
        integration: new WebSocketMockIntegration("MockMediaTypesIntegration"),
      },
      multipleContentTypes: {
        integration: new WebSocketMockIntegration("MockMultipleContentTypesIntegration"),
      },
      withoutOperationIdPost: {
        integration: new WebSocketMockIntegration("MockWithoutOperationIdPostIntegration"),
      },
    };
  }
}
"
`;

exports[`Typescript Async Infrastructure Code Generation Script Unit Tests Generates With single.yaml 3`] = `
"export * from "./api";
export * from "./functions";
export * from "./mock-integrations";"
`;
